source/fuzz/transformation_replace_add_sub_mul_with_carrying_extended.cpp

```diff
@@ -0,0 +1,227 @@
+          ->GetSingleWordOperand(kOpTypeIndexSignedness);
+  switch (instruction_opcode) {
+    case SpvOpIAdd:
+    case SpvOpISub:
+      return operand_1_signedness == 0 && operand_2_signedness == 0;
+    default:
+      return true;
+  }
+}
+
+void TransformationReplaceAddSubMulWithCarryingExtended::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  // |message_.struct_fresh_id| must be fresh.
+  assert(fuzzerutil::IsFreshId(ir_context, message_.struct_fresh_id()) &&
+         "|message_.struct_fresh_id| must be fresh");
+
+  // |message_.result_id| must refer to an OpIAdd or OpISub or OpIMul
+  // instruction
+  auto original_instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.result_id());
+  auto original_instruction_opcode = original_instruction->opcode();
+  assert((original_instruction_opcode == SpvOpIAdd ||
+          original_instruction_opcode == SpvOpISub ||
+          original_instruction_opcode == SpvOpIMul) &&
+         "The instruction must have the opcode: OpIAdd or OpISub or OpIMul");
+
+  // Both operands must be signed or unsigned depending on the instruction.
+  uint32_t operand_1_type_id =
+      ir_context->get_def_use_mgr()
+          ->GetDef(original_instruction->GetSingleWordOperand(
+              kArithmeticInstructionIndexLeftOperand))
+          ->type_id();
+
+  uint32_t operand_2_type_id =
+      ir_context->get_def_use_mgr()
+          ->GetDef(original_instruction->GetSingleWordOperand(
+              kArithmeticInstructionIndexRightOperand))
+          ->type_id();
+
+  assert(operand_1_type_id == operand_2_type_id &&
+         "The type ids of components must be equal");
+
+  uint32_t operand_1_signedness =
+      ir_context->get_def_use_mgr()
+          ->GetDef(operand_1_type_id)
+          ->GetSingleWordOperand(kOpTypeIndexSignedness);
+  uint32_t operand_2_signedness =
+      ir_context->get_def_use_mgr()
+          ->GetDef(operand_2_type_id)
+          ->GetSingleWordOperand(kOpTypeIndexSignedness);
+
+  switch (original_instruction_opcode) {
+    case SpvOpIAdd:
+    case SpvOpISub:
+      assert(operand_1_signedness == 0 && operand_2_signedness == 0 &&
+             "Components must be unsigned.");
+      break;
+    default:
+      break;
+  }
+
+  fuzzerutil::UpdateModuleIdBound(ir_context, message_.struct_fresh_id());
+
+  // Insert the OpCompositeExtract. This instruction takes the first component
+  // of the result which represents low-order bits of the operation.
+  auto instruction_composite_extract =
+      original_instruction->InsertBefore(MakeUnique<opt::Instruction>(
+          ir_context, SpvOpCompositeExtract, operand_1_type_id,
+          message_.result_id(),
+          opt::Instruction::OperandList(
+              {{SPV_OPERAND_TYPE_ID, {message_.struct_fresh_id()}},
+               {SPV_OPERAND_TYPE_LITERAL_INTEGER,
+                {kOpCompositeExtractIndexLowOrderBits}}})));
+
+  switch (original_instruction_opcode) {
```

Comment:
I didn't fully check, but I *think* you could replace all this with:

SpvOpcode new_instruction_opcode;
switch (original_instruction_opcode) {
  // work out what the new opcode is depending on the existing opcode and the signedness, if relevant.
  // assert false in a default case, and assign new_instruction_opcode to SpvOpMax in that case (to keep compilers happy)
}

instruction_composite_extract->InsertBefore(....); <--- A single constructor call that takes new_instruction_opcode as an argument.